File: realestate-core/src/main/java/util/Permissions.java
package util;

import java.util.Map;
import java.util.Set;

/**
 * 权限点定义与「角色 → 权限」映射。
 *
 * <p>对应需求报告 R-001。设计原则：写操作放开，删除操作收回。
 * AGENT 持有 9 项中的 6 项，三个删除权限（房屋 / 客户 / 带看）被收回。
 */
public final class Permissions {

    // ------------------------------------------------------------ 权限点

    public static final String HOUSE_VIEW = "house:view";
    public static final String HOUSE_ADD = "house:add";
    public static final String HOUSE_DELETE = "house:delete";

    public static final String CUSTOMER_VIEW = "customer:view";
    public static final String CUSTOMER_ADD = "customer:add";
    public static final String CUSTOMER_DELETE = "customer:delete";

    public static final String VIEWING_VIEW = "viewing:view";
    public static final String VIEWING_ADD = "viewing:add";
    public static final String VIEWING_DELETE = "viewing:delete";

    // -------------------------------------------------------------- 角色

    public static final String ROLE_ADMIN = "ADMIN";
    public static final String ROLE_AGENT = "AGENT";

    private static final Set<String> ADMIN_PERMISSIONS = Set.of(
            HOUSE_VIEW, HOUSE_ADD, HOUSE_DELETE,
            CUSTOMER_VIEW, CUSTOMER_ADD, CUSTOMER_DELETE,
            VIEWING_VIEW, VIEWING_ADD, VIEWING_DELETE);

    /** AGENT 可增可查，但不能删除任何数据 */
    private static final Set<String> AGENT_PERMISSIONS = Set.of(
            HOUSE_VIEW, HOUSE_ADD,
            CUSTOMER_VIEW, CUSTOMER_ADD,
            VIEWING_VIEW, VIEWING_ADD);

    private static final Map<String, Set<String>> ROLE_PERMISSIONS = Map.of(
            ROLE_ADMIN, ADMIN_PERMISSIONS,
            ROLE_AGENT, AGENT_PERMISSIONS);

    private Permissions() {
    }

    /** 角色对应的权限集合。未知角色返回空集合（默认拒绝，见 Session.can） */
    public static Set<String> of(String role) {
        if (role == null) {
            return Set.of();
        }
        return ROLE_PERMISSIONS.getOrDefault(role.trim().toUpperCase(), Set.of());
    }

    /** 角色字符串是否可识别 */
    public static boolean isKnownRole(String role) {
        return role != null && ROLE_PERMISSIONS.containsKey(role.trim().toUpperCase());
    }

    /** 角色的中文显示名，用于界面展示 */
    public static String displayName(String role) {
        if (role == null) {
            return "未登录";
        }
        switch (role.trim().toUpperCase()) {
            case ROLE_ADMIN:
                return "管理员";
            case ROLE_AGENT:
                return "经纪人";
            default:
                return "未知角色";
        }
    }
}

File: frontend/src/utils/permissions.js
/**
 * 权限点与角色的前端侧定义。对应需求报告 R-001 与 G-024。
 *
 * 这些字符串必须与 realestate-core/src/main/java/util/Permissions.java 一致 ——
 * 但请记住：**前端拿权限只决定「显示什么」，真正的拦截在后端**
 * （core 的 controller 里那道 Session.can(...)）。
 * 界面上藏起一个按钮从来不算安全措施，这条规矩从 R-001 起就没变过。
 */

export const PERMISSION = {
  HOUSE_VIEW: 'house:view',
  HOUSE_ADD: 'house:add',
  HOUSE_DELETE: 'house:delete',
  CUSTOMER_VIEW: 'customer:view',
  CUSTOMER_ADD: 'customer:add',
  CUSTOMER_DELETE: 'customer:delete',
  VIEWING_VIEW: 'viewing:view',
  VIEWING_ADD: 'viewing:add',
  VIEWING_DELETE: 'viewing:delete'
}

export const ROLE = {
  ADMIN: 'ADMIN',
  AGENT: 'AGENT'
}

/** 房屋状态。与 core 的 model/House.java 保持一致 */
export const HOUSE_STATUS = {
  VACANT: '空置',
  RENTED: '已租出'
}

export const HOUSE_STATUS_OPTIONS = [
  { label: '全部状态', value: '' },
  { label: HOUSE_STATUS.VACANT, value: HOUSE_STATUS.VACANT },
  { label: HOUSE_STATUS.RENTED, value: HOUSE_STATUS.RENTED }
]

/**
 * 表单里可选的状态（不含筛选用的那个空选项）。
 *
 * 取值必须与 core 的 model/House.java 里 STATUSES 完全一致——后端有
 * isValidStatus 再校验一次，前端这里只是把可选项摆出来。
 */
export const HOUSE_STATUS_CHOICES = [
  { label: HOUSE_STATUS.VACANT, value: HOUSE_STATUS.VACANT },
  { label: HOUSE_STATUS.RENTED, value: HOUSE_STATUS.RENTED }
]

/** 房东下拉里表示「不是选已有房东，而是录入一个新的」的哨兵值 */
export const NEW_LANDLORD = '__new_landlord__'
